server.rb

```diff
@@ -2,26 +2,62 @@
 require "sinatra/activerecord"
 require 'pry'
 require 'sinatra/flash'
-enable :sessions
 require_relative 'app/models/user'
 set :views, Proc.new { File.join(root, "app/views") }
 
+configure do
+  enable :sessions
+  set :session_secret, "secret"
+end
+
+VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-]+(\.[a-z]+)*\.[a-z]+\z/i
+
 get '/' do
-  erb :sign_in
+  if !logged_in?
+    erb :sign_in
+  else
+    redirect to '/dashboard'
+  end
 end
 
 get '/dashboard' do
   erb :dashboard
 end
 
 post '/session/new' do
+
+  if VALID_EMAIL_REGEX.match(params[:username])
+    user = User.find_by(:email => params[:username])
+  else
+    user = User.find_by(:username => params[:username])
+  end
```

Comment:
You could offload some of the logic to the database by doing something like...

```
user = User.find_by("username = % or email = %", params[:username], params[:username])
```

... if you want to do raw SQL in non-standard ActiveRecord use cases. But you would **NEED** to use the `%` in the string and pass the variables in as subsequent arguments, because this forces ActiveRecord to do some sanitization and prevent SQL injection. ([This](http://gavinmiller.io/2015/fixing-sql-injection-vulnerabilities/) is a really good discussion).

But honestly, your code might be longer and more explicit, but that's not bad at all - it makes it clear what is going on and what the distinction is, and it doesn't even make SQL injection a possibility, like if some other developer goes in and changes...

```
User.find_by("email = %", params[:username])
```

... to...

```
User.find_by("email = #{params[:username]}")
```

... because they think it looks cleaner or more Ruby-ish.